### ISSUE TITLE
Use OAuth instead of collecting user passwords directly

### ISSUE THREAD

**Issue by Timsonrobl** [Franiac/TwitchLeecher#714]
There is a MAJOR difference between asking to authenticate Twitch account and asking to **enter password directly** outside of trusted browser.
This raises serious security concerns. Please adopt similar techniques as Chatty and other local apps that uses OAuth authorization flow to do this operation securely.

Also this opens all kinds of opportunities for people to make fake apps to steal user authentication data caused by this decision and can be devastating for the community.

**Comment by Franiac** [comment]
TL uses the official OAtuh authentication workflow in order to retrieve an user access token (on application startup) as pointed out [HERE](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth/#oauth-authorization-code-flow). The token gets used to communicate with the API.

There is no user access token without entering the username and password. There are app access tokens as described [HERE](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth/#oauth-client-credentials-flow). But these tokens would make all TL users share the same 800 API call limit per minute. And as stated in the documentation, it is not meant for client applications like TL:

> As mentioned earlier, app access tokens are only for server-to-server API requests

If you are talking about the sub-only login where basically the login form of the Tiwtch website gets shown, that is different. But again, it is the official login form of the  Twitch.tv website. If you think that is unsafe, go ahead and tell Twitch, not me.

TL uses an integrated Chromium browser engine, which is the exact same browser engine as Goggle Chrome uses with the exact same security and sandbox features. So I gon't get where your security concerns originate from. Sounds like a lot of made up paranoia, and/or you do not understand what you are actually talking about.

> Also this opens all kinds of opportunities for people to make fake apps to steal user authentication data caused by this decision and can be devastating for the community.

What the hell are you even talking about? Because TL now has a login form it is somehow prone to be faked? Dude...
To sum this up... you make a lot of claims but not a single on is actually backed up by any sort of proof or reason why you make that exact claim. And the claim that TL does not use OAuth tokens is simply wrong from the getgo.

TL is not unsecure or does fishy stuff. It only uses official endpoints documented by Twitch in order to retreive OAuth tokens. Login forms are displayed in an official build of the Chromium engine and sandboxed exactly like a "real" browser would be. So I give you one more chance to explain yourself and the paranoia.

**Comment by Timsonrobl** [comment]
@Franiac
You probably misunderstood me. I'm sure this app uses Twitch API to get OAuth tokens. The problem is *how* it receives the token.
The whole point of OAuth authorization flow is to keep user authentication data (i.e. passwords and OTP codes) completely separate from third parties while enabling them to use serviced requiring authentication. 

> OAuth (Open Authorization) is an open standard for access delegation, commonly used as a way for Internet users to grant websites or applications access to their information on other websites but **_without giving them the passwords_**.

(from wiki)

I'm not accusing TL of doing fishy stuff, I'm talking about violation the principal security approach of OAuth. Integrated Chromium is not a secure browser because it is under complete control of third party app maker. You can break into any sandboxes and inject arbitrary code to any pages rendered in webview. Therefore this approach requires a **complete trust** between app maker and user. And this is fundamentally wrong. We've seen many examples when legitimate apps and browser extensions were hijacked / sold to fishy companies. Users should never trust their passwords to third parties period.
Login form that looks like a Twitch logging form (and probably is in your TL case) IS prone to be faked. An app looking exactly like TL can be easily made to steal user authentication data, and previously that were impossible.
I'm urging you to at lease keep this discussion open so others can chime in. Being concerned about entering passwords to third party apps is not paranoia.

**Comment by Timsonrobl** [comment]
On an unrelated note: currently [yt-dlp/yt-dlp](https://github.com/yt-dlp/yt-dlp) is able to download any videos from twitch without user login (I assume they parse pages and user GraphQL directly with anonymous sessions). Are the reasons that can't be done in TL purely legal?

**Comment by Franiac** [comment]
I did not misunderstand you.

> OAuth (Open Authorization) is an open standard for access delegation, commonly used as a way for Internet users to grant websites or applications access to their information on other websites but without giving them the passwords.

This is ONE usecase. But Twitch does not use OAuth this way for the Helix API (as in password-less tokens). You are confusing OAuth and login credentials. OAuth is a workflow. It does not define what data is used for authentication. The Twitch Helix API NEEDS a user token received by entering username and password. TL uses the documented workflow of Twitch. The Twich website uses password-less tokens if you are not logged in, but at the same time such a token cannot access the Helix API.

[READ THIS](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth/#oauth-authorization-code-flow)

This is how Twitch wants you to do it, and that is exactly what TL does. You cannot access the Helix API without logging in as an user in a client application.

> Therefore this approach requires complete trust between app maker and user. And this is fundamentally wrong.

This is absolute nonsens. How do you think Steam works for example? Steam is an application with an integrated browser where you have to login. Do you trust Valve? Spotify is an application with an integrated browser. Do you trust Spotify? TL is an application with an integraten browser. Do you trust me?

The funny thing is, Steam and Spotify are even closed source, TL is open source. You can actually check, what happens. So my claim is, it is even more secure than Steam and Spotify because you can actually check, what happens with your username and password.

> Login form that looks like a Twitch login form (and probably is in your TL case) IS prone to be faked. An app looking exactly like TL can be easily made to steal user authentication data, and previously that was impossible.

There are so many things wrong in this statement that I don't even want to try correcting it all. I can implement a fake app that looks exactly like Spotify... and now what? Every application can be faked.... so? How is that a problem only for TL just because now it has a login form?

> I'm urging you to at least keep this discussion open so others can chime in. Being concerned about entering passwords to third-party apps is not paranoia.

Definitely not, because the amount of wrong information in your posts is way too high. You have not yet understood how Twitch wants you to access the Helix API even with the documentation provided, so please refrain from providing any more "tips" on how do do it correctly. The API is not the website and the OAuth workflow used for the website is different to the OAuth workflow of the API. You are mixing up so many things.

> I assume they parse pages and use GraphQL directly with anonymous sessions

So in your world, parsing the website source and using the private GraphQL API is better than following the official Twitch documentation on how to access their public API? Big oof.

Funny point: TL is one of the few apps that actually use the API the intended way. GQL should not be used, it is not public.

And that's definitely the end of this discussion. Please refrain from opening up another issue on this matter.

**Comment by Franiac** [comment]
Just in case this gets more attention than necessary I want to quickly address two important things, just without the paranoia part.

You should not input your username and password in third-party software when it's not open source and the flow where your credentials go cannot be tracked. TL is open source and the room for malicious behavior is minimal due to the fact that everyone can see the source code. In TL the username and password are input in the login form displayed by the in-app browser. Yes, I as a developer COULD break into the browser sandbox and collect your secret data. Fact is, I don't. And that again is backed up by the source code of TL. So your username and password of your Tiwtch account NEVER leave the browser sandbox and do not even flow through the TL source code at all. The only time your credentials get processed is on the Twitch server, resulting in the API authentication token. If I changed this behaviour, people would burn this repository down to the ground faster than I could delete it 😄 

What the OP expects is that TL opens up the system default browser in order to input your credentials there. This way the user can be absolutely sure, that his credentials do not touch the third-party source code and get collected for malicious activity. Like I stated, correct for closed-source software but not TL that hundereds of developers have eyes on.

The claim that the pages and therefor the login forms shown by TL can be faked is quite weird. Many game launchers (World of Tanks, Steam, Ubisoft Connect, GOG Galaxy,...) are desktop applications with integrated browsers (usually the Chromium engine). So they work exactly the same way as TL does, also providing login forms. If someone can fake something in TL, they can fake the login dialog of World of Tanks or Steam since the browser technology is the same. The argument is, it is easier to fake something in an in-app browser than an actual browser. This is correct due to the fact that the in-app browser can also be accessed by the source code of the app. If it was actually an issue, none of the desktop applications like Steam or Spotify could be trusted.

And in general, applications can be faked. TL can be faked, true. But this is an odd argument. I can fake Spotify and upload it. If you download my fake Spotify app and use it, Spotify cannot do anything about it. So that's not an TL issue.

So basically the OP is not wrong, but also not right. The username/password issue the OP has is entirely negated by the open-source characteristic of TL. As long as you download TL from this repository, you are in a very safe place. Even if you don't trust me, trust the source code... that never lies.
